src/test/java/software/amazon/nio/spi/s3/config/S3NioSpiConfigurationTest.java

```diff
@@ -36,24 +43,164 @@ public void setup(){
+        then(config.getMaxFragmentSize()).isEqualTo(S3_SPI_READ_MAX_FRAGMENT_SIZE_DEFAULT);
+        then(config.getEndpointProtocol()).isEqualTo("https");
+        then(config.getEndpoint()).isEmpty();
+        then(config.getRegion()).isNull();
+        then(config.getCredentials()).isNull();
+    }
+
+    @Test
+    public void overridesAsMap() {
+        assertThrows(NullPointerException.class, () -> new S3NioSpiConfiguration((Map)null));
+
+        Map<String, String> map = new HashMap<>();
+        map.put(S3_SPI_READ_MAX_FRAGMENT_SIZE_PROPERTY, "1212");
+        S3NioSpiConfiguration c = new S3NioSpiConfiguration(map);
+
+        assertEquals(1212, c.getMaxFragmentSize());
     }
 
     @Test
     public void getS3SpiReadMaxFragmentSize() {
-        assertEquals(Integer.parseInt(S3_SPI_READ_MAX_FRAGMENT_SIZE_DEFAULT), config.getMaxFragmentSize());
+        assertEquals(S3_SPI_READ_MAX_FRAGMENT_SIZE_DEFAULT, config.getMaxFragmentSize());
 
         assertEquals(1111, overriddenConfig.getMaxFragmentSize());
-        assertEquals(Integer.parseInt(S3_SPI_READ_MAX_FRAGMENT_SIZE_DEFAULT), badOverriddenConfig.getMaxFragmentSize());
+        assertEquals(S3_SPI_READ_MAX_FRAGMENT_SIZE_DEFAULT, badOverriddenConfig.getMaxFragmentSize());
     }
 
     @Test
     public void getS3SpiReadMaxFragmentNumber() {
-        assertEquals(Integer.parseInt(S3_SPI_READ_MAX_FRAGMENT_NUMBER_DEFAULT), config.getMaxFragmentNumber());
+        assertEquals(S3_SPI_READ_MAX_FRAGMENT_NUMBER_DEFAULT, config.getMaxFragmentNumber());
 
         assertEquals(2, overriddenConfig.getMaxFragmentNumber());
-        assertEquals(Integer.parseInt(S3_SPI_READ_MAX_FRAGMENT_NUMBER_DEFAULT), badOverriddenConfig.getMaxFragmentNumber());
+        assertEquals(S3_SPI_READ_MAX_FRAGMENT_NUMBER_DEFAULT, badOverriddenConfig.getMaxFragmentNumber());
+    }
+
+    @Test
+    public void withAndGetRegion() {
+        then(new S3NioSpiConfiguration().getRegion()).isNull();
+
+        Properties env = new Properties();
+        env.setProperty(AWS_REGION_PROPERTY, "region1");
+
+        final S3NioSpiConfiguration C = new S3NioSpiConfiguration(env);
+        then(C.getRegion()).isEqualTo("region1");
+        then(C.withRegion("\tregion2 ")).isSameAs(C); then(C.getRegion()).isEqualTo("region2");
+        then(C.withRegion(" \t ").getRegion()).isNull();
+        then(C.withRegion("").getRegion()).isNull();
+        then(C.withRegion(null).getRegion()).isNull();
+    }
+
+    @Test
+    public void withAndGetEndpoint() {
+        then(config.withEndpoint("somewhere.com:8000")).isSameAs(config);
+        then(config.getEndpoint()).isEqualTo("somewhere.com:8000");
+        then(config.withEndpoint(" somewhere.com:8080\t").getEndpoint()).isEqualTo("somewhere.com:8080");
+        then(config.withEndpoint("   ").getEndpoint()).isEqualTo("");
+        then(config.withEndpoint(null).getEndpoint()).isEqualTo("");
+
+        try {
+           config.withEndpoint("noport.somewhere.com");
```

Comment:
This pattern should be allowed. Port numbers are not usually present in official AWS service endpoints which look like `https://<service>.us-east-1.amazonaws.com`

Reply:
done